src/main/java/org/csanchez/jenkins/plugins/kubernetes/AllContainersRunningPodWatcher.java

```diff
@@ -0,0 +1,142 @@
+package org.csanchez.jenkins.plugins.kubernetes;
+
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import io.fabric8.kubernetes.api.model.ContainerStatus;
+import io.fabric8.kubernetes.api.model.Pod;
+import io.fabric8.kubernetes.api.model.PodStatus;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.KubernetesClientException;
+import io.fabric8.kubernetes.client.KubernetesClientTimeoutException;
+import io.fabric8.kubernetes.client.Watcher;
+
+/**
+ * A pod watcher reporting when all containers are running
+ */
+public class AllContainersRunningPodWatcher implements Watcher<Pod> {
+    private static final Logger LOGGER = Logger.getLogger(AllContainersRunningPodWatcher.class.getName());
+
+    private final CountDownLatch latch = new CountDownLatch(1);
+    private final AtomicReference<Pod> reference = new AtomicReference<>();
+
+    private Pod pod;
+
+    private KubernetesClient client;
+
+    private PodStatus podStatus;
+
+    public AllContainersRunningPodWatcher(KubernetesClient client, Pod pod) {
+        this.client = client;
+        this.pod = pod;
+        this.podStatus = pod.getStatus();
+        if (areAllContainersRunning(pod)) {
+            reference.set(pod);
+            latch.countDown();
+        }
+    }
+
+    @Override
+    public void eventReceived(Action action, Pod pod) {
+        LOGGER.log(Level.FINEST, "[{0}] {1}", new Object[]{action, pod.getMetadata().getName()});
+        this.podStatus = pod.getStatus();
+        switch (action) {
+            case MODIFIED:
+                if (areAllContainersRunning(pod)) {
```

Comment:
do you need to check for terminated pods? otherwise if pod fails you will wait until the timeout instead of failing early

Reply:
done